### PR TITLE
bpo-32166: Drop Python 3.4 code from asyncio

### PR DESCRIPTION
**Pull request by asvetlov**
`types.coroutine`, `types.CoroutineType`, `inspect.iscoroutinefunction` and `collections.abc.Coroutine` / `collections.abc.Awaitable` are always present.

https://bugs.python.org/issue32166

### CODE REVIEW

**Review by 1st1 — APPROVED** [review]

File: Lib/asyncio/coroutines.py [hunk 1]
```diff
@@ -39,49 +35,12 @@ def _is_debug_mode():
...
-        yield from gen
-    value = (1, 2, 3)
-    gen = MyGen()
-    coro = yield_from_gen(gen)
-    next(coro)
-    coro.send(value)
-    return gen.send_args != (value,)
-_YIELD_FROM_BUG = has_yield_from_bug()
-del has_yield_from_bug
+_types_coroutine = types.coroutine
```
Comment (serhiy-storchaka) [inline Lib/asyncio/coroutines.py]: Wouldn't the code look clearer if inline these definitions?

File: Lib/asyncio/coroutines.py [hunk 2]
```diff
@@ -259,17 +202,12 @@ def wrapper(*args, **kwds):
...
-            _inspect_iscoroutinefunction(func))
+    return (_inspect_iscoroutinefunction(func) or
+            getattr(func, '_is_coroutine', None) is _is_coroutine)
 
 
-_COROUTINE_TYPES = (types.GeneratorType, CoroWrapper)
-if _CoroutineABC is not None:
-    _COROUTINE_TYPES += (_CoroutineABC,)
-if _types_CoroutineType is not None:
-    # Prioritize native coroutine check to speed-up
```
Comment (serhiy-storchaka) [inline Lib/asyncio/coroutines.py:270]: Keep this comment.

**Review by serhiy-storchaka — COMMENTED** [review]
There is also a code related to Python < 3.5.

File: Lib/asyncio/coroutines.py (hunk 1 above)
Comment (asvetlov) [reply Lib/asyncio/coroutines.py]: Done

File: Lib/asyncio/coroutines.py (hunk 2 above)
Comment (asvetlov) [reply Lib/asyncio/coroutines.py:270]: Make sense

**Comment by asvetlov** [comment]
> There is also a code related to Python < 3.5.

If you are talking about `yield from` support -- I think we should keep it for sake of backward compatibility.

**Comment by serhiy-storchaka** [comment]
Aren't these changes break backward compatibility with Python < 3.5?

I think it is worth to open an issue on the tracker for discussion.

File: Lib/asyncio/coroutines.py [hunk 3]
```diff
@@ -259,17 +194,13 @@ def wrapper(*args, **kwds):
...
-    # asyncio.iscoroutine.
-    _COROUTINE_TYPES = (_types_CoroutineType,) + _COROUTINE_TYPES
+ # Prioritize native coroutine check to speed-up
```
Comment (serhiy-storchaka) [inline Lib/asyncio/coroutines.py]: The second line of the comment has been lost.

File: Lib/asyncio/coroutines.py [hunk 4]
```diff
@@ -218,23 +156,20 @@ def coro(*args, **kw):
             if (base_futures.isfuture(res) or inspect.isgenerator(res) or
                 isinstance(res, CoroWrapper)):
                 res = yield from res
-            elif _AwaitableABC is not None:
+            elif Awaitable is not None:
```
Comment (serhiy-storchaka) [inline Lib/asyncio/coroutines.py]: Is `Awaitable` imported?

Comment (asvetlov) [reply Lib/asyncio/coroutines.py]: Good catch. Fixed.

**Comment by asvetlov** [comment]
After PR merging asyncio code cannot be executed on Python 3.4, sure.
But
```
async def f():
    ...
yield from f()
```
and
```
@asyncio.coroutine
def g():
    ...
await g()
```
is still supported.
By backward compatibility I mean an ability to execute third party libraries written with python 3.4 support (aiohttp 2.x for example) on Python 3.7.

**Comment by asvetlov** [comment]
@serhiy-storchaka I created https://bugs.python.org/issue32166 as you suggested.

**Comment by asvetlov** [comment]
@1st1 should the PR be backported to 3.6?
The change is safe but it is not a bugfix.

**Comment by 1st1** [comment]
No. 3.6 is freezed for any refactorings—this policy ensures we dont break something accidentally. Let's treat 3.6 as a legacy code :)

File: Lib/asyncio/coroutines.py [hunk 5]
```diff
@@ -9,16 +9,14 @@
...
+from collections.abc import Awaitable, Coroutine
+
 from . import constants
 from . import events
 from . import base_futures
 from .log import logger
 
 
-# Opcode of "yield from" instruction
-_YIELD_FROM = opcode.opmap['YIELD_FROM']
```
Comment (serhiy-storchaka) [inline Lib/asyncio/coroutines.py:19]: Is opcode still used?

File: Lib/asyncio/coroutines.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/asyncio/coroutines.py]: Isn't `Awaitable is not None` always true?

File: Lib/asyncio/coroutines.py [hunk 6]
```diff
@@ -259,17 +196,14 @@ def wrapper(*args, **kwds):
...
-_COROUTINE_TYPES = (types.GeneratorType, CoroWrapper)
-if _CoroutineABC is not None:
-    _COROUTINE_TYPES += (_CoroutineABC,)
-if _types_CoroutineType is not None:
-    # Prioritize native coroutine check to speed-up
-    # asyncio.iscoroutine.
-    _COROUTINE_TYPES = (_types_CoroutineType,) + _COROUTINE_TYPES
+# Prioritize native coroutine check to speed-up
+# asyncio.iscoroutine.
+_COROUTINE_TYPES = (types.CoroutineType, Coroutine,
```
Comment (serhiy-storchaka) [inline Lib/asyncio/coroutines.py]: In the current code the natural type `types.GeneratorType` is tested before the abstract type `Coroutine`.

File: Lib/asyncio/coroutines.py (hunk 5 above)
Comment (asvetlov) [reply Lib/asyncio/coroutines.py:19]: No. It was used for making a workaround for bug already fixed in Python 3.4

File: Lib/asyncio/coroutines.py (hunk 6 above)
Comment (asvetlov) [reply Lib/asyncio/coroutines.py]: When async/await is used `types.CoroutineType` and `collections.abc.Coroutine` are more frequent than old styled generators.
That's why I've changed the order.

**Comment by asvetlov** [comment]
Redundant check for `Awaitable is not None` dropped. Thanks for pointing out.

File: Lib/asyncio/coroutines.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/asyncio/coroutines.py]: But checking an instance of `types.GeneratorType` is much faster than running the Python code in `collections.abc.Coroutine.__subclasshook__`.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM except that I'm not sure about the order in `_COROUTINE_TYPES`.

**Comment by asvetlov** [comment]
Well, if check for generator is much faster than ABC -- let's put it first.

File: Lib/asyncio/unix_events.py [hunk 7]
```diff
@@ -680,7 +646,7 @@ def _start(self, args, shell, stdin, stdout, stderr, bufsize, **kwargs):
             # needed by close_fds=False on Python 3.3 and older
             # (Python 3.4 implements the PEP 446, socketpair returns
             # non-inheritable sockets)
-            _set_inheritable(stdin_w.fileno(), False)
+            os.set_inheritable(stdin_w.fileno(), False)
```
Comment (asvetlov) [inline Lib/asyncio/unix_events.py]: > Does not the above comment make this unneeded? @vstinner, what are your thoughts?

Moved the question from #4633
@vstinner please review

Comment (vstinner) [reply Lib/asyncio/unix_events.py]: os.set_inheritable(stdin_w.fileno(), False) is useless on Python 3.4 and newer. Remove the call with its comment.

Comment (vstinner) [reply Lib/asyncio/unix_events.py]: If you don't trust me, please trust the tests ;-)
https://github.com/python/cpython/blob/ef83806f5ec01f34f6cbf4ebb5752875b5961f7e/Lib/test/test_socket.py#L5145-L5146

**Comment by vstinner** [comment]
> @1st1 should the PR be backported to 3.6?

NO! Such refactoring change must not be backport to prevent any risk of regression.

File: Lib/asyncio/unix_events.py (hunk 7 above)
Comment (asvetlov) [reply Lib/asyncio/unix_events.py]: Thanks.
The call is removed.

**Review by serhiy-storchaka — APPROVED** [review]

**Review by vstinner — APPROVED** [review]
LGTM. It's nice to see such large cleanup changes, to *simplify* asyncio code base! It seems like Python core became better ;-)

**Comment by asvetlov** [comment]
Thanks everybody for review!

**Pull request merged**